README.md
# Epictope

Software for predicting epitope tag insertion sites in proteins.

Epictope is an R package/pipeline to identify epitope tag insertion sites for proteins of interest. It uses four features of protein structure; sequence conservation, secondary structure, disordered binding regions, and relative solvent accessabilty to predict suitable internal locations for tag insertion.

The primary score for Epictope relies on a "least-worst" approach, where insertion site suitability is identified by  positions where the lowest scoring feature is the highest. For a given position, we sort the feature scores from lowest to highest, and take the lowest score. We then plot the lowest score for each position, and determine positions where this lowest score is highest to be suitable positions for tagging. 

This repository contains the code source of the R Epictope package, step-by-step R Markdown and Jupyter notebooks to run the complete workflow, wrapped scripts for simplified workflow executation, and instructions to adjust the weight and effect of each considered feature. The package requires local installations of BLAST, MUSCLE, and DSSP to run (those can be installed as a package with Epictope). You will need at least 3GB of disk space.

## Table of Contents
- [Methodology](#methodology)
- [Installation](#installation)
- [Usage](#usage)
- [Examples](#examples)

## Methodology

<figure style="display: inline-block; text-align: center;">
  <img src="images/workflow.png" alt="Alt text" title="Epictope workflow." width="75%">
  <figcaption>Epictope workflow. Starting from a protein of interest, the Epictope workflow automates multiple sequence alignment, downloading of predicted protein structure from AlphaFold2, and fetching disordered binding regions from IUPred2A.</figcaption>.
</figure>


### Sequence conservation

Sequence conservation is used to guide internal epitope-tagging approaches. Regions of relatively low conservation are unlikely to be involved in the critical function of the protein. To identify these regions for a protein of interest, we first BLAST the query protein against the proteomes of a diverse set of model organisms. By default, we compare the query sequence against the proteomes of _Mus musculus_ (mouse), _Bos taurus_ (cow), _Canis lupus familiaris_ (dog), Gallus gallus (chicken), _Homo sapiens_ (human), _Takifugu rubripes_ (pufferfish), and _Xenopus tropicalis_ (western clawed frog). Using BLAST, we identify the highest scoring match in each organism, sorted by the lowest E-value. We then align the retrieved sequences with the query protein using MUSCLE, a multiple sequence alignment program, and calculate the shannon entropy at each position. We use [Shannon entropy](https://en.wikipedia.org/wiki/Entropy_(information_theory)) as a simple measure of the calculate the variability of amino acids at each position in the alignment. A lower Shannon entropy indicates low variability, or high sequence conservation at the position, and it should therefore be avoided for tag insertion. Conversly, a high Shannon entropy indicates a relatively low degree of sequence conservation, and potential suitability for tagging.


<figure style="display: inline-block; text-align: center;">
  <img src="images/msa.png" alt="Alt text" title="Tcf21 Multiple Sequence Alignment." width="75%">
  <figcaption>Example Multiple Sequence Alignment for Tcf21 protein sequences from position 1 to 70. The protein of interest or query is the tcf21 protein from Danio rerio (zebrafish). Amino acids identical between all species are in red, non-identical between at least one species in blue, and gaps are highlighted in yellow. From this alignment, red regions would be unfavorable to tag insertion.</figcaption>.
</figure>

### Solvent accessibility

 Relative Solvent Accessibility (RSA) is a measure of the surface area of a folded protein that is accessible to a solvent, typically the cytoplasmic fluid. It is calculated by dividing the solvent accessible surface area (SASA) of an amino acid by the maximum possible solvent accessible surface area for that residue. SASA values are assigned with Define Secondary Structure of Proteins (DSSP). The DSSP program defines secondary structure, geometrical features and solvent exposure of proteins, given atomic coordinates in Protein Data Bank (PDB) format. Values used for the maximum possible solvent accessible surface area were taken from [this study](https://doi.org/10.1371/journal.pone.0080635). We use the [Alphafold2 predicted structure from the European Bioinformatics Institute (EBI)](https://alphafold.ebi.ac.uk/) as the source PDB for DSSP calculations.


### Secondary structure

Secondary structure is the local spatial conformation of the polypeptide backbone for the protein of interest. Certain structures, such as alpha helices or beta sheets, are more defined and disruption of these structure is likely to affect protein structure. As with solvent accessibility, we use DSSP to define the secondary structure of the protein from its PDB file. By default, we assign helices (GHI) and sheets (E) feature scores of 0. Hydrogen bonded turns (T), residues in isolated Beta bridges (B), and bends (S) scores of 0.5, and coils scores of 1. For all features, higher values indicate greater suitability for tag insertion. 

### Disordered binding 

Disordered binding regions are sections of a protein that do not have a well-defined structure on their own, but can undergo a disorder-to-order transition when they bind to specific protein partners. To avoid these regions, we use [ANCHOR2](https://iupred2a.elte.hu/), a tool that analyzes an amino acid sequence and returns a score of intrinsic disorder depending on a model of the estimated energy potential for residue interactions. To maintain consistency with other features, the disordered binding feature score is taken as 1 minus the ANCHOR2 score.

## Installation

### System requirements
Installing Epictope and its dependencies will require at least 3Gb of disk space. Users should also be familar with using conda, a package manager for macOS/linux and Windows. Conda does not need to be used if users already have access to installations of BLAST, MUSCLE, and DSSP, either locally or on an HPC environment.  

### Software dependencies

To calculate the multiple sequence alignment and secondary characteristics, Epictope relies on local installs of BLAST, MUSCLE, and DSSP. These packages can be installed using Conda, an open-source package management system and environment management system that runs on Windows, macOS, and Linux. Conda installers can be found at the Anaconda [website](https://www.anaconda.com/). Once installed, you may run the follow commands to install the requisite packages. These commands will create a conda environment named Epictope, and install the requisite packages into that environment. Installing Epictope and its dependencies will require at least 3GB of disk space.  


### macOS/Linux installation

For macOS/Linux, commands are issued at the terminal. Dependencies can be installed using the following commands.

1. Download and place the contents of the "install/mac_linux" folder into your project directory. In your terminal, type "ls" to verify the files are in the correct folder.

```bash
curl -o "epictope_install.sh" "https://raw.githubusercontent.com/henrichung/epitope_tag/main/install/mac_linux/epictope_install.sh"
ls
```

2. Run the installation scripts in the terminal with the following commands. 
```
chmod +x epictope_install.sh
bash -i epictope_install.sh
```

Additional installation methods for Linux can be found in the Detailed Installation for Linux [page](https://github.com/henrichung/epitope_tag/wiki/Detailed-Linux-Instructions)


### Windows installation

BLAST and MUSCLE are not available for installation on Windows with conda and have to be installed separately. We provide a wrapper script to install these programs and the conda environment. 

Sure, here are the detailed steps to open Anaconda Prompt, create a new project folder, and then 'cd' (change directory) into it on a Windows machine:


1. In Anaconda prompt, download and place the "epictope_install.bat" folder into your project directory with curl. Type "dir" to verify the files are in the correct folder.
```bash
curl -o epictope_install.bat https://raw.githubusercontent.com/henrichung/epitope_tag/main/install/windows/epictope_install.bat
dir
```

2. Run the installation scripts in Anaconda Prompt with the following commands. If prompted by Conda, type "Y" for Yes and then press "Enter".  Click "Yes" if a pop-up window asks if you allow this app to make changes to your advice.
```
epictope_install.bat
```

More detailed instructions for Windows can be found in the Detailed Installation for Windows [page](https://github.com/henrichung/epitope_tag/wiki/Detailed-Windows-Instructions)

## Usage

Here, we provide usage examples to demonstrate how to use Epictope. Each example includes a brief description and code snippets or commands to showcase the function. These examples assume the installation steps have been followed.

### Example 1A: Generating epictope predictions on macOS/Linux

For our example, we investigate the Smad5 gene for Zebrafish. Searching for the protein transcript in [Uniprot](https://www.uniprot.org/uniprotkb/Q9W7E7/entry), we find it's UniprotID is "Q9W7E7"

Run the Epictope workflow with the following commands in the terminal.
```bash
conda activate epictope
Rscript install.R
Rscript single_score.R Q9W7E7
```

### Example 1B: Generating epictope predictions on Windows

On windows, the commands are the same as for Linux, except Windows uses a backwards slash "\\" instead of a forward slash "/".
Run the Epictope workflow with the following commands in Anaconda Prompt.
```bash
conda activate epictope
Rscript install.R
Rscript single_score.R Q9W7E7
```

### Example 2: Viewing your results.

The Epictope workflow generates a "\<UniprotID\>_score.csv" file (ex: Q9W7E7_score.csv), containing the individual feature scores for each position, the minimum score across features for each position, and a weighted sum score of all features. These values can be plotted in the data visualization tool of choice.

For convenience, we provide a "plot_scores.R" scripts that generates a plot of the minimum score for each position in the sequence using a rolling average of window size 7.

The plot script can be run in the same way as previous commands.

```bash
Rscript plot_scores.R outputs/Q9W7E7_score.csv
"outputs/Q97W7E7.tiff"
```

<figure style="display: inline-block; text-align: center;">
  <img src="images/Q9W7E7_score.png" alt="Alt text" title="Tcf21 Multiple Sequence Alignment." width="50%">
  <figcaption>Example minimum score plot for Q9W7E7. Values smoothed over a windows size of 7.</figcaption>
</figure>

### Workflow notebooks
Example workflows with the Epictope package are available in the **vignettes** folder. Workflows are available as both [R Markdown Documents](https://rmarkdown.rstudio.com/) and [Jupyter](https://jupyter.org/) notebooks. These workflows go through the Epictope workflow step by step in an interactive session or an IDE. *IDE usage requires access to local installations of BLAST, MUSCLE, and DSSP by the IDE.*

### Macro scripts
The scripts `install.R` and `single_score.R` are provided in the **scripts** folder of this repo to enable one-command operation.
To run, download the `install.R` and `single_score.R` scripts from this repository either directly from the github page or using git clone.

- [install.R](https://github.com/henrichung/epitope_tag/blob/main/scripts/install.R)
  - This script first downloads the proteomes for the species used in the multiple sequence alignment from the NCBI FTP page.
  - It then converts these sequences into usable files for BLAST.
  - This file need to be re-run if the user changes the species considered in the multiple sequence alignment.

- [single_score.R](https://github.com/henrichung/epitope_tag/blob/main/scripts/single_score.R)
  - This script takes a UniprotID as input and performs the Epictope workflow for that protein.
  - It first retrieves the amino acid sequence and Alphafold2 predicted structure for the protein.
  - It then BLASTs the protein against the proteomes of the animals used in the multiple sequence alignment, retrieves the highest scoring match (score measured by the lowest E-value), and aligns the matched proteins along with the query in a multiple sequence alignment.
  - It then determines the secondary structure, solvent accessibility, and disordered binding regions for the protein.
  - It combines all feature scores into a summary dataframe.
  - The dataframe annotates each residue position with its feature scores and final tagging score.
  - This file is saved to an /outputs folder with the name of the protein followed by '_score.csv'.
  - For example, the protein used in the examples saves a "outputs/P57102_score.csv" file.


From the terminal, these scripts can be run as follows.

```
Rscript install.R 
Rscript single_score.R "P57102" # replace 'P57102' with the UniprotID for your protein of interest.
```

Each script can also be opened in an IDE such as Rstudio, and run interactively line by line.

### User configuration

A second scoring function used by Epictope sums the calculated scores for the protein features, with equal weight assigned to secondary structure, disordered binding regions, and solvent accessibility. Sequence conservation carries by default carries a higher weight, at 1.5 times that of the other features.

Users can adjust the weight of each feature by modifying the "config_defaults.R" file. This file allows fine-tuning of parameters in Epictope, including the weight of each feature, defining the species used in the multiple sequence alignment, scoring tag suitability for secondary structures, and determining maximum solvent accessibility values.

Epictope searches for a "config.R" file in the working directory. If it doesn't find one, it will utilize default values. In the scripts folder, an example "config_defaults.R" value is provided. To use, edit and rename the file to "config.R" and place it anywhere in your project directory. 


## License 

Epictope is distributed open-source under the GPL3 license.


NEWS.md
# epictope 0.0.1

* Added a `NEWS.md` file to track changes to the package.
* Initial release of package.


code/seqplot.R
# Load required packages
library(tidyverse)
library(ggplot2)
library(gridExtra)

# Read in score files
score_folder <- "outputs/scores"
files = list.files(score_folder, "*.tsv")
score_data <- list()
for( i in 1:length(files)){
	score_data[[i]] = read_tsv(paste(score_folder, files[i], sep = "/"))
}
names(score_data) <- sapply(files, function(.x){strsplit(.x, "\\.")[[1]][1]})

# plot score files
fun_plot <- function(.x, title = ""){
	require(gridExtra)
	require(ggplot2)
	temp <- .x
	p1 <- ggplot(temp, aes(x = id, y = normalized_entropy)) + 
		geom_bar(stat = "identity", aes(fill = "Shannon Entropy")) + 
		ylab("") + xlab("") + ylim(0, 1.5) + 
		scale_x_continuous(breaks=temp$id, labels=temp$residue) +
		scale_fill_manual(name = "", breaks = "Shannon Entropy", values = "darkolivegreen4") +
		theme(legend.position="top")

	p2 <- ggplot(temp, aes(x = id, y = struct_no)) + 
		geom_bar(stat = "identity", aes(fill = "Predicted Structure Interference"))  + 
		ylab("") + xlab("") + ylim(0, 1) + 
		scale_x_continuous(breaks=temp$id, labels=temp$residue) +
		scale_fill_manual(name = "", breaks = "Predicted Structure Interference", values = "#E69F00") +
		theme(legend.position="top")

	p3 <- ggplot(temp, aes(x = id, y = soluble_sa)) + 
		geom_bar(stat = "identity", aes(fill = "Soluble Surface Area")) + 
		ylab("") + xlab("") + ylim(0, 1) + 
		scale_x_continuous(breaks=temp$id, labels=temp$residue) +
		scale_fill_manual(name = "", breaks = "Soluble Surface Area", values = "#56B4E9") +
		theme(legend.position="top")
	
	p4 <- ggplot(temp, aes(x = id, y = inv_anchor2)) +
		geom_bar(stat = "identity", aes(fill = "Protein Binding Site")) + 
		ylab("") + xlab("") + ylim(0, 1) + 
		scale_x_continuous(breaks=temp$id, labels=temp$residue) +
		scale_fill_manual(name = "", breaks = "Protein Binding Site", values = "darkorchid4") +
		theme(legend.position="top")

	p5 <- temp %>%
		mutate(smooth_score = zoo::rollmean(score, 11, fill = NA)) %>%
		ggplot(aes(x = id, y = score)) +
		geom_bar(stat = "identity", aes(fill = "Tagging Score")) + 
		geom_line(aes(y = smooth_score)) + 
		ylab("") + xlab("") + 
		scale_x_continuous(breaks=temp$id, labels=temp$residue) +
		scale_fill_manual(name = "", breaks = "Tagging Score", values = "gray") +
		theme(legend.position="top")

	p6	<- temp %>%
		select(residue, normalized_entropy, struct_no, soluble_sa, inv_anchor2, score) %>%
		rowid_to_column("id") %>%
		group_by(id, residue) %>%
		pivot_longer(cols = c("normalized_entropy", "struct_no", "soluble_sa", "inv_anchor2", "score"), names_to = "var", values_to = "vals") %>%
		slice(which.min(vals)) %>%
		ungroup() %>%
		mutate(var = factor(var, levels = c("normalized_entropy", "struct_no", "soluble_sa", "inv_anchor2"))) %>%
		mutate(vals = ifelse(vals == 0, 0.05, vals)) %>%
		ggplot(aes(x = id, y = vals, fill = var)) +
		geom_bar(stat = "identity") +
		ylab("") + xlab("") + ylim(0, 1) + 
		scale_x_continuous(breaks=temp$id, labels=temp$residue) +
		scale_fill_manual(name = "Variable", values = c("darkolivegreen4", "#E69F00", "#56B4E9", "darkorchid4")) +
		theme(legend.position="top")

	res <- grid.arrange(p1, p2, p3, p4, p5, p6, nrow = 6, top = title, bottom = "Higher values are better") 
	return(res)
}

for( i in 1:length(score_data)){
	title = names(score_data)[i]
	a <- fun_plot(score_data[[i]], title = title)
	pdf(paste("figures/", title, ".pdf", sep = ""), height = 12, width = 40)
	plot(a)
	dev.off()
}

# check correlations
all_scores <- bind_rows(score_data, .id = "tf") %>%
	select(c("soluble_sa", "iupred2", "anchor2", "entropy", "struct_no", score)) %>%
	rowwise() %>%
	mutate(inv_anchor2 = 1-anchor2) %>%
	rowwise() %>%
	select(-anchor2) %>%
	filter(!is.na(iupred2)) %>%
	select(inv_anchor2, everything()) 

# calculate kendall correlation
method = "spearman"
colnames(all_scores) <- c("Inverse Anchor2 Score", "Soluble Surface Area", "IUPred2", "Shannon Entropy", "Secondary Structure Score", "Tagging Score")
res <- cor(all_scores, method = method, use = "complete.obs")
ind <- upper.tri(res, diag = FALSE)
res[ind] <- 0


#a <- all_scores$`Shannon Entropy`
#b <- all_scores$`Inverse Anchor2 Score`
#cor(a, b, method = method, use = "complete.obs")
# reshape correlation data to long
res_long <- res %>%
	reshape2::melt() %>%
	filter(value != 0 & Var1 != "Tagging Score") %>%
	arrange(desc(value)) %>%
	rowwise() %>%
	mutate(value2 = signif(value, 2)) %>%
	rowwise()

p1 <- res_long %>%
	ggplot(aes(x = Var2, y = Var1, fill = value)) + 
	geom_tile() +
	geom_text(aes(label=value2)) +
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name=paste(method, "Correlation", sep = "\n")) + xlab("") + ylab("") +
 scale_x_discrete(limits = rev(levels(res_long$Var2))[2:6]) + 
 theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5)) +
 labs(title = "Correlation of Tagging Variables") 
pdf(paste("corr", method, ".pdf", sep = ""), height = 6, width = 8)
p1
dev.off()

#####
res_spearman <- cor(all_scores, method = "spearman")
ind_spearman <- upper.tri(res_spearman, diag = TRUE)
res_spearman[ind_spearman] <- 0

# reshape correlation data to long
res_long_spearman <- res_spearman %>%
	reshape2::melt() %>%
	filter(value != 0 & Var1 != "Tagging Score") %>%
	arrange(desc(value)) %>%
	rowwise() %>%
	mutate(value2 = signif(value, 2)) %>%
	rowwise()

p1_spearman <- res_long_spearman %>%
	ggplot(aes(x = Var1, y = Var2, fill = value)) + 
	geom_tile() +
	geom_text(aes(label=value2)) +
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name="Spearman\nCorrelation") + xlab("") + ylab("") +
 theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5)) +
 labs(title = "Correlation of Tagging Variables") 
pdf("corr_spearman.pdf", height = 6, width = 8)
p1_spearman
dev.off()

#####
p2 <- all_scores %>%
	reshape2::melt(id = "Shannon Entropy") %>%
	filter(variable != "Secondary Structure Score") %>%
	filter(variable != "Tagging Score") %>%
	ggplot(aes(x = `Shannon Entropy`, y = value)) + 
	geom_point(size = 0.5, alpha = 0.1) +
	xlab("Shannon Entropy") + ylab("") + 
	labs(title = "Scatterplot of Shannon Entropy with Variables") +
	facet_wrap(~variable, scales = "free")
pdf("entropy.pdf", height = 4, width = 12)
p2
dev.off()

p3 <- all_scores %>%
	reshape2::melt(id = "Soluble Surface Area") %>%
	filter(variable != "Secondary Structure Score") %>%
	filter(variable != "Tagging Score") %>%
	ggplot(aes(x = `Soluble Surface Area`, y = value)) + 
	geom_point(size = 0.5, alpha = 0.1) +
	xlab("Soluble Surface Area") + ylab("") + 
	labs(title = "Scatterplot of Soluble Surface Area with Variables") +
	facet_wrap(~variable, scales = "free")
pdf("solubility.pdf", height = 4, width = 12)
p3
dev.off()

p4 <- all_scores %>%
	reshape2::melt(id = "Inverse Anchor2 Score" ) %>%
	filter(variable != "Secondary Structure Score") %>%
	filter(variable != "Tagging Score") %>%
	ggplot(aes(x = `Inverse Anchor2 Score`, y = value)) + 
	geom_point(size = 0.5, alpha = 0.1) +
	xlab("Inverse Anchor2 Score" ) + ylab("") + 
	labs(title = "Scatterplot of Inverse Anchor2 Score with Variables") +
	facet_wrap(~variable, scales = "free")
pdf("inverse_anchor2.pdf", height = 4, width = 12)
p4
dev.off()

p5 <- all_scores %>%
	reshape2::melt(id = "IUPred2") %>%
	filter(variable != "Secondary Structure Score") %>%
	filter(variable != "Tagging Score") %>%
	ggplot(aes(x = `IUPred2`, y = value)) + 
	geom_point(size = 0.5, alpha = 0.1) +
	xlab("IUPred2") + ylab("") + 
	labs(title = "Scatterplot of IUPred2 with Variables") +
	facet_wrap(~variable, scales = "free")
pdf("iupred2.pdf", height = 4, width = 12)
p5
dev.off()

p6 <- all_scores %>%
	reshape2::melt(id = "Secondary Structure Score") %>%
	filter(variable != "Tagging Score") %>%
	mutate(`Secondary Structure Score` = as.factor(`Secondary Structure Score`)) %>%
	ggplot(aes(x = `Secondary Structure Score`, y = value)) + 
	geom_violin() + 
	geom_jitter(size = 0.1, alpha = 0.1) +
	xlab("Secondary Structure Score") + ylab("") + 
	labs(title = "Scatterplot of Secondary Structure Score with Variables") +
	facet_wrap(~variable, scales = "free")
pdf("sec_struct.pdf")
p6
dev.off()

#### 

